core/embed/rust/src/ui/shape/algo/blur.rs

```diff
@@ -0,0 +1,357 @@
+use crate::ui::geometry::Offset;
+/// This is a simple and fast blurring algorithm that uses a box filter -
+/// a square kernel with all coefficients set to 1.
+///
+/// The `BlurFilter` structure holds the context of a simple 2D window averaging
+/// filter - a sliding window and the sum of all rows in the sliding window.
+///
+/// The `BlurFilter` implements only five public functions - `new`, `push`,
+/// `push_read`, `pop` and `pop_ready`.
+///
+/// The `new()` function creates a blur filter context.
+///   - The `size` argument specifies the size of the blurred area.
+///   - The `radius` argument specifies the length of the kernel side.
+///
+/// ```rust
+/// let blur = BlurFilter::new(size, radius);
+/// ```
+///
+/// The `push_ready()` function returns the row from the source bitmap
+/// needed to be pushed
+///
+/// The `push()` function pushes source row data into the sliding window and
+/// performs all necessary calculations.
+///
+/// ```rust
+/// if let Some(y) = blur.push_ready() {
+///     blur.push(&src_bitmap.row(y)[x0..x1]);
+/// }
+/// ```
+///
+/// The `pop_ready()` function returns the row from the destination bitmap
+/// that can be popped out
+///
+/// The `pop()` function pops the blurred row from the sliding window.
+///
+/// ```rust
+/// if let Some(y) = blur.pop_ready() {
+///     blur.pop(&mut dst_bitmap.row(y)[x0..x1]);
+/// }
+/// ```
+use core::mem::size_of;
+
+const MAX_RADIUS: usize = 4;
+const MAX_SIDE: usize = 1 + MAX_RADIUS * 2;
+const MAX_WIDTH: usize = 240;
+
+pub type BlurBuff = [u8; MAX_WIDTH * (MAX_SIDE * 3 + size_of::<u16>() * 3) + 8];
+
+type PixelColor = u16;
+
+#[derive(Default, Copy, Clone)]
+struct Rgb<T> {
+    pub r: T,
+    pub g: T,
+    pub b: T,
+}
+
+impl Rgb<u16> {
+    #[inline(always)]
+    fn mulshift(&self, multiplier: u32, shift: u8) -> Rgb<u8> {
+        Rgb::<u8> {
+            r: ((self.r as u32 * multiplier) >> shift) as u8,
+            g: ((self.g as u32 * multiplier) >> shift) as u8,
+            b: ((self.b as u32 * multiplier) >> shift) as u8,
+        }
+    }
+}
+
+impl From<u16> for Rgb<u16> {
+    #[inline(always)]
+    fn from(value: u16) -> Rgb<u16> {
+        Rgb::<u16> {
+            r: (value >> 8) & 0xF8,
+            g: (value >> 3) & 0xFC,
+            b: (value << 3) & 0xF8,
+        }
+    }
+}
+
```

Comment:
the nice thing to do here is use Self
```suggestion
    #[inline(always)]
    fn from(value: u16) -> Self {
        Self {
            r: (value >> 8) & 0xF8,
            g: (value >> 3) & 0xFC,
            b: (value << 3) & 0xF8,
        }
    }
}

```

Reply:
modified, see dfb196e10e593d7c495ff524e49bcfb5fa28df21